Libraries/Opc.Ua.Server/Diagnostics/CustomNodeManager.cs

```diff
@@ -4621,65 +4629,42 @@ private class CacheEntry
         /// </summary>
         protected NodeState LookupNodeInComponentCache(ISystemContext context, NodeHandle handle)
         {
-            lock (Lock)
-            {
-                if (m_componentCache == null)
-                {
-                    return null;
-                }
-
-                CacheEntry entry = null;
+            bool componentPathExists = !string.IsNullOrEmpty(handle.ComponentPath);
+            NodeId nodeId = componentPathExists ? handle.RootId : handle.NodeId;
 
-                if (!String.IsNullOrEmpty(handle.ComponentPath))
+            if (m_componentCache?.TryGetValue(nodeId, out CacheEntry entry) == true)
+            {
+                if (componentPathExists)
                 {
-                    if (m_componentCache.TryGetValue(handle.RootId, out entry))
-                    {
-                        return entry.Entry.FindChildBySymbolicName(context, handle.ComponentPath);
-                    }
+                    return entry.Entry.FindChildBySymbolicName(context, handle.ComponentPath);
                 }
                 else
                 {
-                    if (m_componentCache.TryGetValue(handle.NodeId, out entry))
-                    {
-                        return entry.Entry;
-                    }
+                    return entry.Entry;
                 }
-
-                return null;
             }
+
+            return null;
         }
 
         /// <summary>
         /// Removes a reference to a component in thecache.
         /// </summary>
         protected void RemoveNodeFromComponentCache(ISystemContext context, NodeHandle handle)
         {
-            lock (Lock)
-            {
-                if (handle == null)
-                {
-                    return;
-                }
+            bool componentPathExists = !string.IsNullOrEmpty(handle.ComponentPath);
+            NodeId nodeId = componentPathExists ? handle.RootId : handle.NodeId;
 
-                if (m_componentCache != null)
+            if (m_componentCache?.TryGetValue(nodeId, out CacheEntry entry) == true)
+            {
+                lock (Lock)
                 {
-                    NodeId nodeId = handle.NodeId;
-
-                    if (!String.IsNullOrEmpty(handle.ComponentPath))
-                    {
-                        nodeId = handle.RootId;
-                    }
-
-                    CacheEntry entry = null;
+                    int refCount = Interlocked.Decrement(ref entry.RefCount);
 
-                    if (m_componentCache.TryGetValue(nodeId, out entry))
+                    if (refCount == 0)
                     {
-                        entry.RefCount--;
-
-                        if (entry.RefCount == 0)
-                        {
-                            m_componentCache.Remove(nodeId);
-                        }
+                        //this will only remove the value if it did not change after retrieving from the dictionary
```

Comment:
does this ever remove, as the refCount of the entry was changed before, if this is not validated not shure if add / remvoe can be achieved thread safe?

Reply:
confirmed working in tests